hazelcast-client/src/main/java/com/hazelcast/client/util/ClientDelegatingFuture.java

```diff
@@ -144,24 +134,22 @@ public V get() throws InterruptedException, ExecutionException {
     @SuppressWarnings({"checkstyle:npathcomplexity", "checkstyle:cyclomaticcomplexity"})
     public V get(long timeout, TimeUnit unit) throws InterruptedException,
             ExecutionException, TimeoutException {
-        if (!done || !isResponseSet()) {
-            synchronized (mutex) {
-                if (!done || !isResponseSet()) {
-                    try {
-                        response = resolveMessageToValue(future.get(timeout, unit));
-                        if (deserializeResponse && deserializedValue == null) {
-                            deserializedValue = serializationService.toObject(response);
-                        }
-                    } catch (InterruptedException e) {
-                        error = e;
-                    } catch (ExecutionException e) {
-                        error = e;
-                    }
-                    done = true;
-                }
+        if (responseMessage.get() == NULL_OBJECT) {
+            Object response;
+            try {
+                response = future.get(timeout, unit);
+            } catch (InterruptedException e) {
+                response = e;
+            } catch (ExecutionException e) {
+                response = e;
             }
+
+            responseMessage.compareAndSet(NULL_OBJECT, response);
```

Comment:
I believe this is broken in case of exceptions. So imagine 2 threads waiting for a future and one thread gets interrupted and he therefor sets the 'responseMessage' to that exception. Now the second thread, which isn't interrupted at all.. will also get this interrupted exception.